src/aes128gcm.rs

```diff
@@ -356,9 +338,174 @@ impl<'a> PlaintextRecord<'a> {
         output[0..ciphertext.len()].copy_from_slice(&ciphertext);
         Ok(ciphertext.len())
     }
+}
 
-    pub(crate) fn encrypted_size(&self) -> usize {
-        self.plaintext.len() + self.padding + ECE_TAG_LENGTH
+/// Iterator returning record-sized chunks of plaintext + padding.
+///
+/// Given a plaintext, an amount of padding data to add, and a target encrypted record
+/// size, this function returns an iterator of `PlaintextRecord` structs such that:
+///
+///    * The encrypted size of each plaintext chunk plus its padding will be equal
+///      to the given record size, except for the final record which may be shorter.
+///
+///    * Each record has at least one padding byte; if necessary, additional padding
+///      bytes will be inserted beyond what was requested by the caller in order
+///      to meet this requirement. (This ensures each record has enough room for the
+///      padding delimiter byte).
+///
+///    * The plaintext is distributed as evenly as possible between records. Records
+///      consisting entirely of padding will only be produced in degenerate cases such
+///      as where the caller requested far more padding than available plaintext, or
+///      where the requested total size falls just beyond a record boundary.
```

Comment:
I'll admit, I couldn't entirely get my head around the previous record-chunking algorithm, but as far as I could tell it was doing a kind of greedy allocation of as much padding a possible to the *start* of the record stream. For example, if you requested 10 bytes of plaintext and 10 bytes of padding split between two records, you'd get 1 plaintext + 9 padding in the first record, then 9 plaintext and 1 padding in the second.

My read of the padding-related comments in the RFC suggests that we should instead by trying to distribute the plaintext *as evenly as possible* between the records, to reduce potential for observable timing differences in the processing of each record. I have tried to implement such here (and I intend to add some tests to demonstrate on concrete examples).

I have also removed an error case where we would refuse to ever generate a record consisting entirely of padding. The RFC does specifically warn against naively generating such records, and the algorithm here will only do so under particularly degenerate cases. But IMHO it may be legit for an application to want to pad out e.g. a very short message to a large fixed length, and it's not helpful for us to prevent it from doing so.

I believe we inherited this error case from the C ece library, but note that Martin's [nodejs implementation](https://github.com/web-push-libs/encrypted-content-encoding/blob/master/nodejs/ece.js) does not seem to error out in this case, so I feel pretty comfortable in removing it.

(It's also true that in the current setup, if a caller did hit this error then *there's nothing they can do about it*, because we don't expose the ability to select the amount of padding; callers might just accidentally land on a plaintext + padding size that produces a padding-only block out of bad luck).